specs/packagemanager/PackageManagement.md

```diff
@@ -0,0 +1,1136 @@
+
+Fabrikam app installing Contoso's Muffin package from an https: source.
+
+```c#
+void Install()
+{
+    var package = new Uri("https://contoson.com/muffin.msix");
+    var packageDeploymentManager = PackageDeploymentManager.GetDefault();
+    var options = new AddPackageOptions();
+    var deploymentResult = await packageDeploymentManager.AddPackageByUriAsync(package, options);
+    try
+    {
+        if (deplymentResult.Status == PackageDeploymentStatus.CompletedSuccess)
+        {
+            Console.WriteLine("OK");
+        }
+        else
+        {
+            Console.WriteLine("Error {}", deploymentResult.ExtendedError());
+        }
+    }
+    catch (Exception ex)
+    {
+        Console.WriteLine(ex.ToString());
+    }
+}
+```
+
+**NOTE:** This differs from the AddPackageAsync() example by the method name
+(`AddPackageByUriAsync()` instead of `AddPackageAsync()`) and the target type (https: source as URI
+instead of filename as string).
+
+## 4.3. AddPackageSetAsync()
+
+Fabrikam app installing Contoso's Muffin and Waffle packages via a PackageSet.
+
+```c#
+var Install()
+{
+    var packageSet = new PackageSet();
+    var muffin = new PackageSetItem();
+    muffin.PackageUri = new Uri("c:\\contoso\\muffin-1.2.3.4.msix");
+    packageSet.Add(muffin);
+    var waffle = new PackageSetItem();
+    waffle.PackageUri = new Uri("https://contoso.com/waffle-2.4.6.8.msix");
+    packageSet.Add(waffle);
+
+    var packageDeploymentManager = PackageDeploymentManager.GetDefault();
+    var options = new AddPackageOptions();
+    var deploymentResult = await packageDeploymentManager.AddPackageByUriAsync(packageSet, options);
+    try
+    {
+        if (deplymentResult.Status == PackageDeploymentStatus.CompletedSuccess)
+        {
+            Console.WriteLine("OK");
+        }
+        else
+        {
+            Console.WriteLine("Error {}", deploymentResult.ExtendedError());
+        }
+    }
+    catch (Exception ex)
+    {
+        Console.WriteLine(ex.ToString());
+    }
+}
+```
+
+**NOTE:** This differs from the AddPackageByUriAsync() example by the method name
+(`AddPackageSetAsync` instead of `AddPackageByUriAsync()`), the target type (`PackageSet` containing
+2 URIs instead of a URI), and installing 2 packages instead of 1.
+
+## 4.4. EnsurePackageSetIsReadyAsync()
+
+Fabrikam app installing Contoso's Muffin and Waffle packages if necessary via a PackageSet.
+
+```c#
+var Install()
+{
+    var packageSet = new PackageSet();
+    var muffin = new PackageSetItem();
+    muffin.PackageFamilyName = "contoso.muffin_1234567890abc";
+    muffin.MinVersion = ToVersion(1, 2, 3, 4);
+    muffin.PackageUri = new Uri("c:\\contoso\\muffin-1.2.3.4.msix");
+    packageSet.Add(muffin);
+    var waffle = new PackageSetItem();
+    waffle.PackageFamilyName = "contoso.waffle_1234567890abc";
+    waffle.MinVersion = ToVersion(2, 4, 6, 8);
+    waffle.PackageUri = new Uri("https://contoso.com/waffle-2.4.6.8.msix");
+    packageSet.Add(waffle);
```

Comment:
How would one know which fields of a PackageSetItem need to be populated for each operation?  Which ones are required, optional, or prohibited?

Reply:
Documentation.

Nothing's prohibited. Propeties are either be used or ignored/NA, depending on the operation.

That's also true for PackageManager e.g. not all properties of AddPackageOptions are relevant for all Add scenarios.